### PR TITLE
docs: add map and map operator documentation

### PR DESCRIPTION
**Pull request by JLDLaughlin**
Fixes https://github.com/MaterializeInc/materialize/issues/4622

The `type-map.svg` looks like:
Screen Shot 2020-12-02 at 10 58 31 AM

This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4920)

### CODE REVIEW

File: doc/user/sql-grammar/sql-grammar.bnf [hunk 1]
```diff
@@ -264,6 +264,8 @@ type_interval_val ::=
   'INTERVAL' "'" time_expr+ "'" ( ( head_time_unit 'TO' )? tail_time_unit )?
 type_jsonb ::=
     "'" json_string "'" '::JSONB'
+type_map ::=
+    "'" map_string "'" '::MAP[text=>value_type]'
```
Comment (sploiselle) [inline doc/user/sql-grammar/sql-grammar.bnf:268]: In the above `value_type` should definitely not be quoted to indicated that it's a variable, similar to `map_string`.

Let's try:

```
"'" map_string "'" '::' 'MAP' '[' 'TEXT' '=>' value_type ']'
```

This breaks everything out into their independently parsed components and hints that `TEXT` could be replaced with other values.

Apologies that the BNF diagrams are inconsistent; we only use them cosmetically, so it's hard to enforce a real structure on them and the standards have drifted over the last year.

File: doc/user/content/sql/types/map.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+## Syntax
+
+{{< diagram "type-map.svg" >}}
+
+Field | Use
+------|-----
+_map&lowbar;string_ | A well-formed map object.
+_value&lowbar;type_ | The [type](../../types) of the map's values.
+
+## Map operators
```
Comment (sploiselle) [inline doc/user/content/sql/types/map.md:30]: This head feels redundant with the following `### Operators`. Looks like it's patterned after other types' "Function + Operator" sections? Until we have functions for it, let's elide the extra heading, and when we add the first function, we can add it in.

File: doc/user/content/sql/types/map.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+You can construct maps from strings using the following syntax:
+```sql
+SELECT '{a=>123.4, b=>111.1}'::map[text=>double] as m;
+```
+```nofmt
+  m
+------------------
+ {a=>123.4,b=>111.1}
+```
+
```
Comment (sploiselle) [inline doc/user/content/sql/types/map.md:55]: errant space

File: doc/user/content/sql/types/map.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+```
+```nofmt
+  nested_map
+------------------
+ {a=>{b=>{c=>d}}}
+```
+
+### Constraints
+
+- `map` keys must be of type [`text`](../text).
```
Comment (sploiselle) [inline doc/user/content/sql/types/map.md:69]: nit: remove `map` from each of these and capitalize the first letter

File: doc/user/content/sql/types/map.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+- `map` values can be of any [type](../../types) as long as the type is uniform.
+- `map` keys must be unique. If duplicate keys are present in a map, only one of the
+  (`key`, `value`) pairs will be retained. There is no guarantee which will be retained.
+
+## Examples
+
+### Operators
+
+#### Field access as `map` (`->`)
```
Comment (sploiselle) [inline doc/user/content/sql/types/map.md:78]: I don't think "Field access" is the right term here; seems like maybe this is "value access" or "Retrieve value with key"?

**Review by sploiselle — APPROVED** [review]
LGTM modulo comments

File: doc/user/sql-grammar/sql-grammar.bnf (hunk 1 above)
Comment (JLDLaughlin) [reply doc/user/sql-grammar/sql-grammar.bnf:268]: That totally makes sense, thanks for your help!

**Pull request merged**